Refactor sweeps_runner - abstract input/output + add Pydantic models for OpTest (sweeps + unit test) for data validation

stevendae opened the pull request:
### Ticket
[pydantic model extension/creation issue](https://github.com/tenstorrent/tt-metal/issues/26031)
[sweep runner refactor issue](https://github.com/tenstorrent/tt-metal/issues/26500)

### Problem description
Sweep test data generation must move towards a production workflow where it can be viewable in Superset. To do so it must be validated by a Pydantic model prior to upload, where the Pydantic model serves as a shared data contract between the generation party and ingestion party. 

### What's changed
Refactored sweep runner to abstract the test source and result destination for removal of redundant code and future extensibility.
Added sweeps-specific Pydantic models to infra/data_collection/pydantic_models.py
Pydantic verification is applied to data generated as a JSON, for elastic and Postgres export it remains untouched (for now).


### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

stevendae committed:
Begin refactor of sweeps_runner.py

stevendae committed:
Complete running tests and publishing tests

stevendae committed:
deserialize for postgres

stevendae committed:
Add normalization and flattening for original_vector_data in result_destination.py

- Introduced `_normalize_original_vector_data` function to convert original_vector_data into a JSON-friendly format.
- Added `_flatten_serialized` function to strip unnecessary wrappers for file output.
- Updated serialization logic in `FileResultDestination` to handle original_vector_data appropriately.
- Ensured device_perf is always included in file output and added error signature handling.

stevendae committed:
Add new Pydantic models for sweeps results and metadata

- Introduced `SweepsResultRecord` to encapsulate individual test vector execution results, including performance metrics and execution status.
- Added `SweepsRunMetadata` to provide high-level metadata for sweeps run sessions.
- Created `SweepsRun` to aggregate run-level metadata and results.
- Updated imports in `sweeps_runner_copy.py` to reflect changes in the serialization logic.

stevendae committed:
Add license headers

stevendae committed:
Refactor sweeps runner and enhance result handling

- Removed the obsolete sweeps_runner_copy.py file.
- Updated sweeps_runner.py to integrate new Pydantic models for result validation.
- Enhanced the export_results method in FileResultDestination to utilize Pydantic for structured data handling.
- Improved argument validation and configuration management in sweeps_runner.py.
- Added logging for better traceability of vector file handling and execution results.

williamlyTT commented on infra/data_collection/pydantic_models.py:
```diff
@@ -373,3 +373,85 @@ class CompleteBenchmarkRun(BaseModel):
+
+    git_author: Optional[str] = Field(None, description="Git author configured in the environment.")
+    git_branch_name: Optional[str] = Field(None, description="Current git branch name.")
+    git_commit_hash: Optional[str] = Field(None, description="Short git commit hash for the run.")
```
Referenced in GHA env as `GITHUB_RUN_ID` or through context variable in the workflows as `${{ github.run_id }}`

stevendae committed:
Update CODEOWNERS and enhance sweep test framework documentation

- Added @stevendae as a code owner for the ttnn-run-sweeps.yaml workflow and the sweep_framework tests.
- Revised the README.md for the sweep test framework to clarify the separation of vector sources and result destinations, including detailed credential requirements for PostgreSQL and Elasticsearch.
- Introduced new options for result destinations, including local JSON export, and updated command examples for running tests with various configurations.

fzhengTT commented on infra/data_collection/pydantic_models.py:
```diff
@@ -373,3 +373,85 @@ class CompleteBenchmarkRun(BaseModel):
+    input_hash: Optional[str] = Field(None, description="Hash of the input vector for deduplication and traceability.")
+
+    # Timing
+    start_time_ts: datetime = Field(description="Timestamp with timezone when the vector execution started.")
+    end_time_ts: datetime = Field(description="Timestamp with timezone when the vector execution ended.")
```
`ts` stands for timestamp, you could call these 2 `sweep_start_ts` and `sweep_end_ts`.

stevendae committed:
Add GitHub pipeline ID support in sweeps metadata and runner

- Introduced `github_pipeline_id` field in `SweepsRunMetadata` to capture the GitHub Actions pipeline run identifier.
- Added `get_github_pipeline_id` function in `sweeps_runner.py` to retrieve the CI pipeline identifier, prioritizing GitHub Actions run ID.
- Updated the `run_sweeps` function to include the GitHub pipeline ID in the run metadata for better traceability.

fzhengTT commented on infra/data_collection/pydantic_models.py:
```diff
@@ -373,3 +373,85 @@ class CompleteBenchmarkRun(BaseModel):
+    e2e_perf: Optional[float] = Field(
+        None, description="End-to-end performance metric for the test (units depend on the specific test)."
+    )
+    device_perf: Optional[Dict[str, Any]] = Field(
```
Will you search for a specific key-value pair from this dict field? If so, we might want to normalize this field into a vertical table in the database for faster querying.
You can do something like this:
```suggestion
    device_perf: Optional[set[DevicePerfMetric]] = Field(None, description="A unique set of device-level performance measurements, when device profiling is enabled.")
```

```
class DevicePerfMetric(BaseModel):
    """
    Metric name and its value.
    """
    metric_name: str = Field(description="Output metric name.")
    metric_value_float: Optional[float] = Field(
        description="Output float metric value."
    )
    class Config:
        frozen = True
```

stevendae replied:
This one is tricky because we haven't thoroughly inspected the perf metrics yet for sweeps. Ability to extract performance was implemented in the test runner therefore it was included. However I do like your proposal of creating a Perf class and set of Perf objects for this field. We can include both the e2e_perf and device_perf in this set.

fzhengTT commented on infra/data_collection/pydantic_models.py:
```diff
@@ -373,3 +373,85 @@ class CompleteBenchmarkRun(BaseModel):
+    user: Optional[str] = Field(None, description="Username associated with the execution environment.")
+
+    # Original vector contents (sanitized/normalized for JSON)
+    original_vector_data: Optional[Dict[str, Any]] = Field(
+        None,
+        description="Original test vector contents captured for traceability, normalized for JSON compatibility.",
+    )
```
same as above for this

stevendae replied:
This field will certainly be searched and queried.

stevendae committed:
Enhance vector source metadata in FileVectorSource and VectorExportSource

- Added `input_hash`, `suite_name`, and `sweep_name` fields to vectors in both FileVectorSource and VectorExportSource classes for improved metadata tracking.
- Updated vector handling logic to ensure these new fields are populated correctly during vector retrieval.

fzhengTT commented on infra/data_collection/pydantic_models.py:
```diff
@@ -373,3 +373,89 @@ class CompleteBenchmarkRun(BaseModel):
+        description="Identifier for the GitHub Actions pipeline run (GITHUB_RUN_ID) or analogous CI pipeline id.",
+    )
+
+    start_time_ts: datetime = Field(description="Timestamp with timezone when the sweeps run started.")
+    end_time_ts: Optional[datetime] = Field(None, description="Timestamp with timezone when the sweeps run ended.")
```
Similarly you can call these 2 `run_start_ts` and `run_end_ts`.

fzhengTT commented on infra/data_collection/pydantic_models.py:
```diff
@@ -373,3 +373,89 @@ class CompleteBenchmarkRun(BaseModel):
+    per-vector execution results.
+    """
+
+    metadata: SweepsRunMetadata = Field(description="Run-level metadata for the sweeps session.")
+    results: List[SweepsResultRecord] = Field(description="List of per-vector sweeps execution results.")
```
We can actually remove this SweepsRun model, then call the model `SweepsRunMetadata` as `SweepsRun`, and put [results](results: List[SweepsResultRecord]) in that model. This way we have one less model to deal with.

stevendae replied:
Sounds good

fzhengTT commented:
A general question: are you still going to use the cicd datasets for data production? Or you'll only use the sweeps models you proposed?

stevendae commented on infra/data_collection/pydantic_models.py:
```diff
@@ -373,3 +373,89 @@ class CompleteBenchmarkRun(BaseModel):
+
+    start_time_ts: datetime = Field(description="Timestamp with timezone when the sweeps run started.")
+    end_time_ts: Optional[datetime] = Field(None, description="Timestamp with timezone when the sweeps run ended.")
+    status: Optional[str] = Field(None, description="Overall run status aggregated from testcases.")
```
enum it is

fzhengTT replied:
does this enum take the same possible values as the above one?

stevendae committed:
Refactor Pydantic models for op test results and enhance result handling

- Renamed and restructured Pydantic models to better represent op test results, including `OpTestResultRecord`, `DevicePerfMetric`, and `OpTestRun`.
- Updated the `FileResultDestination` class to utilize the new models for exporting results, ensuring compatibility with the revised schema.
- Introduced an enumeration for op test statuses to standardize execution status handling.
- Enhanced device performance metrics handling to accommodate the new data structure.

stevendae commented:
> A general question: are you still going to use the cicd datasets for data production? Or you'll only use the sweeps models you proposed?

Based on current approach we will not need to pull cicd datasets for data production. Everything is embedded in our test runners.

stevendae committed:
Refactor Pydantic models and result handling in sweeps framework

- Updated `OpTestResultRecord` and `OpTestRun` models to enforce required fields and improve clarity in execution status and metadata.
- Simplified result destination creation in `sweeps_runner.py` by consolidating configuration handling.
- Added comments for potential future improvements in the `FileResultDestination` class regarding status mapping.

stevendae committed:
Refactor Pydantic models and serialization logic in sweeps framework

- Introduced new `TestStatus` and `RunStatus` enums to standardize test execution statuses.
- Renamed `OpTestResultRecord` to `OpTest` and `DevicePerfMetric` to `PerfMetric` for clarity.
- Updated serialization and deserialization functions to use structured formats, enhancing compatibility with the new models.
- Simplified result handling in `FileResultDestination` to align with the revised Pydantic models.
- Removed obsolete database configuration handling from the parameter generator.

stevendae committed:
Enhance Pydantic models and result handling in data collection

- Introduced `TestVector` model to represent test parameters and their values, allowing for single values, lists, or tuples.
- Updated `FileResultDestination` to utilize the new `TestVector` model for handling original vector data.
- Added functions for flattening nested structures into a single-level dictionary with dotted keys, improving data serialization.
- Refined normalization logic to ensure compatibility with the new model structure.

stevendae committed:
25771/run sweeps workflow (#26814)

### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] [All post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml)
CI passes
- [ ] [Blackhole Post
commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml)
CI with demo tests passes (if applicable)
- [ ] [Model
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml)
CI passes (if applicable)
- [ ] [Device performance
regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml)
CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable) See [recommended dev
flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy
quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml)
CI passes (if applicable)
- [ ] [TG demo tests, for
Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml)
CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml)
CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml)
CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo
tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

stevendae committed:
Merge branch 'produce_op_test_jsons' of github.com:tenstorrent/tt-metal into produce_op_test_jsons

stevendae committed:
Update the pydantic model to reflect the merged version between us and forge as reflected data_airflow repo

stevendae committed:
first iteration of generating data according to the joint schema

stevendae committed:
- Updated `OpRun` model to streamline metadata for sweep runs, removing unnecessary details for clarity.
- Improved comments in `sweeps_runner.py` for better understanding of vector source and result destination setup.
- Enhanced run metadata collection with additional fields for improved tracking and reporting.
- Refined summary counters and logging for better insights into test execution results.

stevendae committed:
Enhance atomic file writing in FileResultDestination to prevent data corruption during interruptions. Implemented temporary file creation and flushing for JSON exports, ensuring valid data integrity. Added logging for available suites in ElasticVectorSource for better visibility.

stevendae committed:
Update README files for tests and sweep framework

- Revised the structure and content of `tests/README.md` to clarify the roles of the Unit Test and Sweep Test frameworks, including updated instructions for generating and running tests.
- Enhanced `tests/sweep_framework/README.md` to direct users to the main README for running the sweeps runner, streamlining the documentation.
- Modified `sweeps_runner.py` to make `--vector-source` and `--result-dest` required arguments, ensuring users specify these critical options when executing tests.

stevendae committed:
Merge branch 'main' into produce_op_test_jsons

stevendae committed:
remove jacob and saad from sweep framework and CI related ownershiper

stevendae committed:
Merge branch 'produce_op_test_jsons' of github.com:tenstorrent/tt-metal into produce_op_test_jsons

stevendae committed:
update readme
